specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,75 @@
 
 ## Overview
 
-// TODO: this is a stub.
+Hooks are a mechanism whereby application developers can add arbitrary behavior
+to flag evaluation. They operate similarly to middleware in many web frameworks.
+
+### Definitions
+
+**Stage**: An explicit portion of the flag evaluation lifecycle. e.g. `before` being "before the evaluation is run.
+**Invocation**: A single call to evaluate a flag. `client.getBooleanValue(..)` is an invocation.
+**API**: The global API singleton.
+
+### Hook context
+
+Hook context exists to provide hooks with information about the invocation.
+
+> Hook context **MUST** provide: the flag key, flag type, evaluation context, default value and a list of  executed hooks by stage.
+
+> Hook context **SHOULD** provide: provider, client
+
+> flag key, flag type, default value properties **MUST** be immutable.
+
+> The evaluation context **MUST** be mutable only within the `before` hook.
+
+### State
+
+> State **MUST** be a map of objects.
+
+> Condition: If your language supports it, state **MUST** be immutable.
+
+### Hook creation and parameters
+
+> Hooks **MUST** specify at least one stage.
+
+> The `before` stage *MUST* run before flag evaluation occurs. It accepts a `hook context` (required) and `state` (optional) as parameters and returns either a `HookContext` or nothing.
+
+```
+HookContext|void before(HookContext, State)
+```
+
+> The `after` stage **MUST** run after flag evaluation occurs. It accepts a `hook context` (required), `flag evaluation details` (required) and `state` (optional). It has no return value.
+
+> The `error` hook runs when errors are encountered in the `before` or `after` state or when flag evaluation errors. It accepts `hook context` (required), `exception` for what went wrong (required), and `state` (optional). It has no return value.
+
+> The `finally` hook **MUST** run after the `before`, `after`, and `error` stages. It accepts a `hook context` (required) and `state` (optional). There is no return value.
+
+> Condition: If `finally` is a reserved word in the language, `afterAll` should be used.
+
+### Hook registration & ordering
+
+> The API, Client and invocation **MUST** have a method for registering hooks which accepts `flag evaluation options`
+
+> Hooks **MUST** be evaluated in the following order: API before, client before, invocation before, invocation after, client after, API after, invocation finally, client finally, API finally. If errors are present, the order is: invocation error, client error, API error.
```

Comment:
To me on initial pass this appears to be a lot of hook stages. Have we actually verified that there is use cases for each one? In my view it'll be cleaner to start with few hook stages and add them as needed vs having a large number upfront and finding it's becoming a maintenance or implementation headache.

Reply:
I see your concern. It think it looks like there's a lot more complexity here than there actually is. The There is really only: `before`, `after`, `error`, and `finally` - it's just that there's 3 places to attach hooks, and so there's a need to define how they interact.

In code, there's not too much complexity here (I hope). From the Application Author's perspective, it's just:

```
OpenFeature.addHooks(new Hook1);

//...

Client client = OpenFeature.getClient();
client.addHooks(new Hook2);

//...

client.getValue('my-flag', 'defaultValue', new Hook3);
```

We are defining the order of these hooks relative to eachother, which isn't too tough in implementation logic either. It looks something like this:

```
// client logic...

// here we are using the correct order (note if it wasn't defined, we'd have to order them somewhow anyway
Hooks[] allHooks = [ ...OpenFeature.getHooks(), ...this.getHooks, ...evaluationOptions.hooks ]

// evaluation all our before hooks
for (h in allHooks) {
  h.before();
}

provider.resolveValue(...);

// evaluation all our after hooks
for (h in allHooks.reverse()) {
  h.after()
}

// more client logic...
```

Let me know if that doesn't address your concern. It might be that there's an easier way to describe this, maybe with a point for list, or with prose. I'm not sure, but I get the point @justinabrahms was trying to make.

Reply:
@toddbaert thanks for the example in code. That does clear it up.

Reply:
You're welcome! If anything, this might mean that when we do the full markdown spec, adding code examples in non-normative auxiliary sections for this requirement is a good idea.